### ISSUE TITLE
clippy.toml ignored if clippy results have been cached

### ISSUE THREAD

**Issue by jimblandy** [rust-lang/rust-clippy#9928]
### Summary

If results from a previous run are present, Clippy ignores changes to a "clippy.toml" file. This makes debugging frustrating, because Clippy's outputs are not a function of its current ostensible inputs.

The following transcript shows that, even if `clippy.toml` has invalid contents, Clippy doesn't notice until `cargo clean` removes its cached results.

```
$ cargo clean
$ cargo clippy
    Checking play-clippy v0.1.0 (/home/jimb/rust/play-clippy)
warning: the `Err`-variant returned from this function is very large
 --> src/main.rs:3:13
  |
3 | fn foo() -> Result<(), Error> {
  |             ^^^^^^^^^^^^^^^^^ the `Err`-variant is at least 4000 bytes
  |
  = note: `#[warn(clippy::result_large_err)]` on by default
  = help: try reducing the size of `Error`, for example by boxing large elements or replacing it with `Box<Error>`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#result_large_err

warning: `play-clippy` (bin "play-clippy") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
$ echo 'msrv = "slurve"' > clippy.toml
$ cargo clippy
warning: the `Err`-variant returned from this function is very large
 --> src/main.rs:3:13
  |
3 | fn foo() -> Result<(), Error> {
  |             ^^^^^^^^^^^^^^^^^ the `Err`-variant is at least 4000 bytes
  |
  = note: `#[warn(clippy::result_large_err)]` on by default
  = help: try reducing the size of `Error`, for example by boxing large elements or replacing it with `Box<Error>`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#result_large_err

warning: `play-clippy` (bin "play-clippy") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
$ cargo clean
$ cargo clippy
    Checking play-clippy v0.1.0 (/home/jimb/rust/play-clippy)
error: error reading Clippy's configuration file. `slurve` is not a valid Rust version

error: could not compile `play-clippy` due to previous error
$ 
```

### Reproducer

I tried this code:

```rust
<code>
```

I expected to see this happen:

Instead, this happened:

**Comment by Alexendoo** [comment]
The situation should be a bit better with #9707, after that changes to existing config files will trigger a rebuild. That'll arrive in the next nightly after https://github.com/rust-lang/rust/pull/104688 is merged

It doesn't fix your example though if `clippy.toml` is newly created, I can't think of a nice way to solve that as cargo will consider the unit changed if a depinfo file is missing

https://github.com/rust-lang/cargo/blob/9865dde6d0b55e4faf4d0dbb10a9c9814955fd1f/src/cargo/core/compiler/fingerprint.rs#L182-L184

**Comment by awused** [comment]
This caused a lot of confusion for me today. I was setting CLIPPY_CONF_DIR to use a new global clippy config file and was very confused why it wasn't working.

More confusingly, clippy will fail instantly if CLIPPY_CONF_DIR is invalid or doesn't contain a clippy.toml/.clippy.toml, but it will not actually read, validate, or use those values until something triggers a rebuild, making me believe the problem was with the lint or the configuration option I was setting.

**Comment by Astralchroma** [comment]
what about `CARGO_MANIFEST_DIR`?

**Comment by Alexendoo** [comment]
We don't track that explicitly, but as that's a variable set by Cargo I don't think we need to. Is there an issue you're running into?

**Comment by Astralchroma** [comment]
Okay

**Comment by tbates-redarc** [comment]
I ran into this issue today, I created a `clippy.toml` file for the first time and was very confused as to why it didn't seem to be working, and then I changed something apparently unrelated and it suddenly started working. Quite a puzzle for new users.
